File: src/decoders/user-parameters.ts
import { AccountKind, type SessionHandle, type UserParameters } from "~/models";
import { decodeUserAuthorizations } from "./user-authorizations";
import { decodeUserResource } from "./user-resource";

export const decodeUserParameters = (parameters: any, session: SessionHandle): UserParameters => {
  let resources: Array<any>;

  switch (session.information.accountKind) {
    case AccountKind.STUDENT:
      resources = [parameters.ressource];
      break;
    case AccountKind.PARENT:
      resources = parameters.ressource.listeRessources;
      break;
    default:
      throw new Error("account kind not supported for now");
  }

  return {
    id: parameters.ressource.N,
    kind: parameters.ressource.G,
    name: parameters.ressource.L,
    resources: resources.map((resource) => decodeUserResource(resource, session)),
    authorizations: decodeUserAuthorizations(parameters.autorisations, parameters.listeOnglets)
  };
};

File: library/src/commonMain/kotlin/decoders/user-parameters.kt
package decoders

import kotlinx.serialization.json.*
import models.AccountKind
import models.InstanceParameters
import models.SessionInformation
import models.UserParameters

fun decodeUserParameters (parameters: JsonObject, sessionInfo: SessionInformation, sessionInstance: InstanceParameters): UserParameters {
    var resources: JsonArray = JsonArray(listOf())
    println(parameters) // TODO

    when (sessionInfo.accountKind) {
        AccountKind.STUDENT -> {}
        AccountKind.TEACHER -> resources = JsonArray(listOf( parameters["ressource"]!!.jsonObject))
        AccountKind.PARENT -> resources = parameters["ressource"]!!.jsonObject["listeRessources"]!!.jsonArray
    }

    return UserParameters(
        id = parameters["ressource"]!!.jsonObject["N"]!!.jsonPrimitive.content,
        kind = parameters["ressource"]!!.jsonObject["G"]!!.jsonPrimitive.int,
        name = parameters["ressource"]!!.jsonObject["L"]!!.jsonPrimitive.content,
        resources = resources.map { decodeUserResource(it.jsonObject, sessionInfo, sessionInstance) },
        authorizations = decodeUserAuthorizations(parameters["autorisations"]!!.jsonObject, parameters["listeOnglets"]!!.jsonArray)
    )
}